### ISSUE TITLE
First thoughts

### ISSUE THREAD

**Issue by VictorBjelkholm** [chriscool/ipfs-sharness-tests#3]
Tried to run sharness with go-ipfs and js-ipfs. First raw thoughts:

  - need to install go-sleep with other deps (missing go-sleep?)
  - install of go-ipfs seems to had failed
    - had to manually copy binary to bin/ipfs
  - js-ipfs support:
    - symlink js-ipfs into bin/ipfs
  - tests have some go-ipfs specific commands (understandable)
  - We need to have a process to solve consensus
    - `version`, basic command but different output on js/go-ipfs. How we make the call which one is right? Version is just cosmetic, others will be based on correctness but "taste" is hard to get consensus on

**Comment by travisperson** [comment]
> tests have some go-ipfs specific commands

Are these actually specific to `go-ipfs`, or are they just commands that `js-ipfs` doesn't currently support? 

> We need to have a process to solve consensus 

I think the expectation would be that any project which wishes to conform to these tests, must follow the format presented. I think things such as the `version` command can be fixed by changing the test to allow for additional characters in some places. For example `ipfs version --all` lists the version as `go-ipfs version: 0.4.16-dev-b3faaad13`. We could allow for the characters up too `ipfs` to be something like `[a-zA-Z]{1,4}-`.

This allows implementations to identify themselves in the verbose format.

> install of go-ipfs seems to had failed
> js-ipfs support

With regards to installing `go-ipfs` and `js-ipfs` support, I think we should make this a bit more generic. A good approach might be to allow the specification of the `ipfs` command, with a fallback to whatever can be found in the `PATH`.

Examples

```bash
$ make deps
```

```bash
$ IPFS_CMD="$HOME/src/github.com/ipfs/cmd/ipfs" make deps
```

```bash
$ IPFS_CMD="$HOME/src/github.com/ipfs/js-ipfs/src/cli/bin.js" make deps
```

We could then create the symlink in the `bin` directory for `ipfs` that all the tests use.

```make
IPFS_DEF != which ipfs
IPFS_CMD ?= $(IPFS_DEF)

$(BIN):
  mkdir -p bin
  cd bin && ln -sf $(IPFS_CMD) ipfs
```

This would enable an easy way to run the sharness tests against any IPFS implementation without having the user mess around too much with their `PATH` themselves, or having implementation specific setups in this project.

It would also allow us in the future to write a small shell script that would mix IPFS implementations, such as running a js-ipfs daemon, and use the go-ipfs cli, which would be an easy way to test the js-ipfs api with a full featured cli.

**Comment by chriscool** [comment]
@VictorBjelkholm @travisperson thanks for your feedback and sorry about replying very late to it.
I have had a hard time managing all the GitHub notifications I get especially from IPFS related projects and your comments were lost for some time in the noise.
 
>  - need to install go-sleep with other deps (missing go-sleep?)

I just made many improvements especially to the Makefile so that the missing tools can be installed and built automatically. See for example: https://github.com/chriscool/ipfs-sharness-tests/commit/b78cdd73b4dd7252ef578e11fd110b8686b413d6

I hope I can finish it today. Otherwise for now one can run a `make test_sharness_deps` in the `go-ipfs/test/sharness` folder of a go-ipfs repo and then copy the files from `go-ipfs/test/bin` into the `bin/` folder of a repo from this project.

  - install of go-ipfs seems to had failed
  - had to manually copy binary to bin/ipfs
  - js-ipfs support:
    - symlink js-ipfs into bin/ipfs

Yeah, for now this project expects people to install go-ipfs or js-ipfs separately first. Then this project expects that a go-ipfs or js-ipfs build has succeeded and that the resulting binary file can be copied or symlinked into the `bin/` folder of this project. Once that is done, then the Sharness test can be run.

**Comment by chriscool** [comment]
>  - tests have some go-ipfs specific commands (understandable)

I haven't looked at that yet. I think the first step would be to make this project work well to test go-ipfs and then adapt things to js-ipfs.

>  - We need to have a process to solve consensus
>    - `version`, basic command but different output on js/go-ipfs. How we make the call which one is right? Version is just cosmetic, others will be based on correctness but "taste" is hard to get consensus on

I already tried to adapt the Sharness tests for js-ipfs (see https://github.com/ipfs/js-ipfs/tree/master/test/sharness) and js-ipfs people are willing to help in this, so I think it will work out and solutions to adapt js-ipfs and adjust Sharness can be quite easily found.

**Comment by chriscool** [comment]
> I think things such as the version command can be fixed by changing the test to allow for additional characters in some places.

Yeah, this is a way to fix things. There are several and it's usually not a big deal.

> With regards to installing go-ipfs and js-ipfs support, I think we should make this a bit more generic. A good approach might be to allow the specification of the ipfs command, with a fallback to whatever can be found in the PATH.

Yeah I also thought about such things and I agree they would be nice to have. I will take a look at these kind of things (after my current improvements are done).
